relay-server/src/http.rs

```diff
@@ -0,0 +1,299 @@
+                }
+                .finish()
+            }
+        }
+    }
+
+    pub fn disable_decompress(&mut self) -> &mut Self {
+        match self {
+            RequestBuilder::Actix(builder) => {
+                builder.disable_decompress();
+            }
+            RequestBuilder::Reqwest { .. } => {
+                // can only be set on client TODO
+            }
+        }
+
+        self
+    }
+
+    pub fn content_encoding(&mut self, encoding: HttpEncoding) -> &mut Self {
+        match self {
+            RequestBuilder::Actix(builder) => {
+                let content_encoding = match encoding {
+                    HttpEncoding::Identity => ContentEncoding::Identity,
+                    HttpEncoding::Deflate => ContentEncoding::Deflate,
+                    HttpEncoding::Gzip => ContentEncoding::Gzip,
+                    HttpEncoding::Br => ContentEncoding::Br,
+                };
+
+                builder.content_encoding(content_encoding);
+            }
+            RequestBuilder::Reqwest {
+                ref mut http_encoding,
+                ..
+            } => {
+                *http_encoding = encoding;
+            }
+        }
+
+        self
+    }
+}
+
+#[allow(clippy::large_enum_variant)]
+pub enum Response {
+    Actix(ClientResponse),
+    Reqwest(reqwest::Response),
+}
+
+impl Response {
+    pub fn status(&self) -> StatusCode {
+        match self {
+            Response::Actix(response) => response.status(),
+            Response::Reqwest(response) => {
+                StatusCode::from_u16(response.status().as_u16()).unwrap()
+            }
+        }
+    }
+
+    pub fn json<T: 'static + DeserializeOwned, E>(
+        self,
+        limit: usize,
+    ) -> Box<dyn Future<Item = T, Error = E>>
+    where
+        E: From<reqwest::Error> + From<actix_web::error::JsonPayloadError> + 'static,
+    {
+        // TODO: apply limit to reqwest
+        match self {
+            Response::Actix(response) => {
+                let future = response.json().limit(limit).map_err(From::from);
+                Box::new(future) as Box<dyn Future<Item = _, Error = _>>
+            }
+            Response::Reqwest(response) => {
+                let future = response.json().boxed_local().compat().map_err(From::from);
+                Box::new(future) as Box<dyn Future<Item = _, Error = _>>
+            }
+        }
+    }
+
+    pub fn consume<E>(self) -> ResponseFuture<Self, E>
+    where
+        E: From<actix_web::error::PayloadError> + 'static,
+    {
+        // consume response bodies to allow connection keep-alive
+        match self {
+            Response::Actix(ref response) => Box::new(
+                response
+                    .payload()
+                    .for_each(|_| Ok(()))
+                    .map(|_| self)
+                    .map_err(From::from),
+            ),
+            Response::Reqwest(_) => {
+                // TODO: is this necessary for reqwest?
```

Comment:
Please verify this. It would be great if reqwest handles this correctly.

Reply:
does not appear to be necessary. how did you test this though?

Reply:
appears not to make a difference in tests

Reply:
We only have a functional test for socket errors, but nothing that asserts that the connection stays open, unfortunately. You would have to manually test this, or figure out a way to assert metrics in tests.

Reply:
I don't think asserting that the connection stays open is sufficient. The way we encountered this in prod (not in dev) was that connections were timing out or abruptly closed in the *next* request.

Reply:
I'm not sure when connections were closed, but they were effectively left in an invalid state after which the downstream had to close and re-establish. Either way, I think we need to verify that reqwest consumes the body or otherwise do this manually.

Reply:
I added request consumption for the reqwest codepath as well since I was not really able to reproduce the problem we had.